### PR TITLE
build2: split out dependencies

### PR DESCRIPTION
**Pull request by r-burns**
Build2's dependency libraries and helper programs
bpkg/bdep are all built and installed separately.

Adds a build2 setup hook for packaging purposes,
inspired by those of cmake and ninja.

Two small patches are applied to build2's source:
one to remove store references from the stored config,
and one to help build2 detect library dirs from NIX_LDFLAGS.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-burns** [comment]
@ofborg eval

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 139347` at 6cdbc72e run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>build2</li>
    <li>libpkgconf</li>
  </ul>
</details>

File: pkgs/development/libraries/libbpkg/default.nix [hunk 1]
```diff
@@ -0,0 +1,48 @@
+{ lib, stdenv
+, build2
+, fetchurl
+, libbutl
+
```
Comment (figsoda) [inline pkgs/development/libraries/libbpkg/default.nix:5]:
```suggestion
```

File: pkgs/development/libraries/libbutl/default.nix [hunk 2]
```diff
@@ -0,0 +1,55 @@
+{ lib, stdenv
+, build2
+, fetchurl
+, libuuid
+
```
Comment (figsoda) [inline pkgs/development/libraries/libbutl/default.nix:5]:
```suggestion
```

File: pkgs/development/tools/build-managers/build2/bdep.nix [hunk 3]
```diff
@@ -0,0 +1,50 @@
+{ lib, stdenv
+, build2
+, fetchurl
+, libbpkg
+, libbutl
+, libodb
+, libodb-sqlite
+
```
Comment (figsoda) [inline pkgs/development/tools/build-managers/build2/bdep.nix:8]:
```suggestion
```

File: pkgs/development/libraries/libodb/default.nix [hunk 4]
```diff
@@ -0,0 +1,46 @@
+{ lib, stdenv
+, build2
+, fetchurl
+
```
Comment (figsoda) [inline pkgs/development/libraries/libodb/default.nix:4]:
```suggestion
```

File: pkgs/development/libraries/libodb-sqlite/default.nix [hunk 5]
```diff
@@ -0,0 +1,57 @@
+{ lib, stdenv
+, build2
+, fetchurl
+, libodb
+, sqlite
+
```
Comment (figsoda) [inline pkgs/development/libraries/libodb-sqlite/default.nix:6]:
```suggestion
```

**Review by figsoda — COMMENTED** [review]
not a review, just noticed some empty lines

File: pkgs/development/libraries/libbpkg/default.nix (hunk 1 above)
Comment (r-burns) [reply pkgs/development/libraries/libbpkg/default.nix:5]: I like an empty line to separate packages from options, but I don't feel strongly about it.

**Comment by risicle** [comment]
I think you're probably going to get asked for separate `init at xyz` commits for each of these packages.

**Comment by r-burns** [comment]
Yeah I can do that. Build2.bootstrap references build2's passthru, so I think build2+build2.bootstrap+libbutl would all have to be added in a single commit at the start for them to pass eval. But all the other packages can be added in separate follup commits if that would be clearer.

**Comment by risicle** [comment]
Confirmed, all builds happily on macos 10.15 & nixos x86_64

**Comment by r-burns** [comment]
cc co-maintainer @vale981

**Comment by vale981** [comment]
lgtm

**Comment by vale981** [comment]
lgtm

**Pull request merged**